packages/util/src/errors.ts

```diff
@@ -103,18 +103,25 @@ export class FirebaseError extends Error {
   }
 }
 
-export class ErrorFactory<ErrorCode extends string> {
+export class ErrorFactory<
+  ErrorCode extends string,
+  ErrorParams extends { readonly [K in ErrorCode]?: ErrorData } = {}
+> {
   constructor(
     private readonly service: string,
     private readonly serviceName: string,
     private readonly errors: ErrorMap<ErrorCode>
   ) {}
 
-  create(code: ErrorCode, data: ErrorData = {}): FirebaseError {
+  create<K extends ErrorCode>(
+    code: K,
+    ...data: K extends keyof ErrorParams ? [ErrorParams[K]] : []
```

Comment:
really nice. 👍

Reply:
Thanks :slightly_smiling_face:

Do you think we can merge this today?